Common/Gfx/Util/D2DBitmapLoader.cpp

```diff
@@ -49,27 +49,37 @@ HRESULT D2DBitmapLoader::LoadBitmapFromFile(const Canvas& canvas, D2DBitmap* bit
 	}
 	bitmap->SetFileTime(fileTime);
 
-	Microsoft::WRL::ComPtr<IWICBitmapDecoder> decoder;
-	Microsoft::WRL::ComPtr<IWICBitmapFrameDecode> decoderFrame;
+	// Microsoft::WRL::ComPtr<IWICBitmapDecoder> decoder;
+	// Microsoft::WRL::ComPtr<IWICBitmapFrameDecode> decoderFrame;
 	Microsoft::WRL::ComPtr<IWICBitmapSource> source;
 
-	HRESULT hr = Canvas::c_WICFactory->CreateDecoderFromFileHandle(
-		(ULONG_PTR)fileHandle,
-		nullptr,
-		WICDecodeMetadataCacheOnDemand,
-		decoder.GetAddressOf());
-	if (SUCCEEDED(hr))
-	{
-		hr = decoder->GetFrame(0U, decoderFrame.GetAddressOf());
-		if (SUCCEEDED(hr))
-		{
-			hr = ConvertToD2DFormat(decoderFrame.Get(), source);
-		}
-	}
-	if (FAILED(hr)) return cleanup(hr);
-
-	const int orientation = GetExifOrientation(decoderFrame.Get());
-	bitmap->SetOrientation(orientation);
+	// HRESULT hr = Canvas::c_WICFactory->CreateDecoderFromFileHandle(
+	// 	(ULONG_PTR)fileHandle,
+	// 	nullptr,
+	// 	WICDecodeMetadataCacheOnDemand,
+	// 	decoder.GetAddressOf());
+	// if (SUCCEEDED(hr))
+	// {
+	// 	hr = decoder->GetFrame(0U, decoderFrame.GetAddressOf());
+	// 	if (SUCCEEDED(hr))
+	// 	{
+	// 		hr = ConvertToD2DFormat(decoderFrame.Get(), source);
+	// 	}
+	// }
+	// if (FAILED(hr)) return cleanup(hr);
+	//
+	// const int orientation = GetExifOrientation(decoderFrame.Get());
+	// bitmap->SetOrientation(orientation);
+
+	// PluginImageData im should be passed as an arg to this function
+	PluginImageData im{};
+	IWICBitmap* iwicBitmap;
+	HRESULT hr = Gfx::Canvas::c_WICFactory->CreateBitmapFromMemory(
+		im.width, im.height,
+		PLUGIN_PIXEL_FORMAT,
+		STRIDE /* I don't know, what should be here, google it */, im.width * im.height, im.pixels, &iwicBitmap);

```

Comment:
for `PLUGIN_PIXEL_FORMAT` i used `GUID_WICPixelFormat32bppBGRA`
for `STRIDE` -> `imageData.width * 4`
for `imageData.pixels` -> arg of type `dword *` is incompatible with parameter of type `byte` (unsigned char)

```cpp
struct PluginImageData
{
DWORD* pixels;
DWORD width;
DWORD height;
uint64_t timestamp;
};
```

Is that how i should make the struct?
Also i still had no idea where to put it, i added it in both `RainmeterAPI.h` and `D2DBitmapLoader` (yep, declared twice).
The second declaration is in order to pass it as an arg to `LoadImageFromPluginData`, i tried 
using void ptr but it was a successful failure :P

Reply:
> for `imageData.pixels` -> arg of type `dword *` is incompatible with parameter of type `byte` (unsigned char)

Well, I guess such APIs are why reinterpret_case exists.

> Is that how i should make the struct?

Yeah, more or less.
I guess, Rainmeter devs would prefer if you used some windows proprietary analogue for `uint64_t` instead of this standardized type.

> Also i still had no idea where to put it, i added it in both `RainmeterAPI.h` and `D2DBitmapLoader` (yep, declared twice).

This struct must be accessible for plugins, so it must be located in a header that plugins include.
As to how to transfer it to the d2d-related classes... Well, it's just a matter of preference. You could create a duplicate class and copy the contents; you could pass it through 3 args (pixels, width, height, you don't need timestamp in d2d); you could include the API header in the d2d file. It's your choice (and maybe a choice of rainmeter devs, if they don't like what you chose), I don't know what will be easier to read and support.

Reply:
>Well, I guess such APIs are why reinterpret_case exists.

I mean `imageData.pixels` should be a pointer to an array that contains the actual pixels (afaik).
And after some thinking, i guess `byte` parameter should be `(unsigned char *)(img.w * img.h * 4 * sizeof(unsigned char))`. I'll try it out and tell you if it works.

But one last note.. you... i mean i think of `imageData.pixels` as an array of `uint8_t`s, but you told me before that it's bad to use this type to represent colors, what should i do in this case?

>Well, it's just a matter of preference.

Oh, i thought it may cause errors.

Reply:
> it's bad to use this type to represent colors

It's absolutely terrible type for doing _math_ on colors.
It's also very _confusing_ for defining colors.
It's a perfectly valid type for _presentation_ on screen, at least as long as OS APIs for showing images on screen use it. There isn't really any alternative, except to use advanced low-level APIs (which Rainmeter is kinda already doing, by the way, but it's still difficult and unnecessary).

Reply:
And I guess that Rainmeter _is_ doing math on it's images.
Who knows if it uses floating pointer textured for intermediate results.

Actually, your point about using a better type for representing is very interesting.
If there is some `PIXEL_FORMAT` that supports `float32`, for example, then you could very well technically make an API for plugins that will use `float32` for transfer.
And _for my plugin_ this would actually be more convenient. Because I already use `float32` for internal calculations, and exporting colors as `uint8_t` is additional work.
However, it may not be as convenient for other plugins. For example, I would imagine that decompressing the image, like you do, yields you integer results, and transforming them into `float32` would require additional work for you, wouldn't bring any benefits and increase memory consumption for each image 4x.

Reply:
Also, if rainmeter actually uses integer textures for intermediate results, then the post above is 100% meaningless.

Reply:
hmm, i thought about creating 2 structs, or 2 pointers at least, one for `float32`, and one for `uint8_t`, and in `D2DBitmapLoader` we check which array is filled, and use the appropriate implementation.

Well, checking which array is filled is not enough (i mean `if float32pixels != nullptr`), because for some reason, what if both arrays are filled? so i thought about adding a variable for the data used, like `if var(a char?) ='float32'` then use that array, else use float32 array.

What do you think of the spaghetti above?

>Also, if rainmeter actually uses integer textures for intermediate results, then the post above is 100% meaningless.

don't say that, you gave me useful infos :(

Reply:
I don't think it would be that useful. I would suppose that Rainmeter devs wouldn't like such unnecessary complication.
Also, considering Rainmeter tries to be memory efficient, I would say that it most probably uses uint8_t textures anyway.

Also, even if it doesn't use uint8_t, for some strange reason, then one transformation `float → int → float` wouldn't make much difference. The issues are in color definitions on user side and in repeated transformations during calculations.
I would suggest you to save your time on this.

Reply:
>And after some thinking, i guess byte parameter should be (unsigned char *)(img.w * img.h * 4 * sizeof(unsigned char)). I'll try it out and tell you if it works.

omg what am i saying lmaoo, after a bit of digging i realized what i should do:

```cpp
HRESULT hr = Gfx::Canvas::c_WICFactory->CreateBitmapFromMemory(
	imageData.width, imageData.height,
	GUID_WICPixelFormat32bppBGRA,
	imageData.width * 4 , (imageData.width * imageData.height * 4) * sizeof(unsigned char), /*(unsigned char*)imageData.pixels*/ reinterpret_cast<BYTE*>(imageData.pixels), &iwicBitmap);	
```

I didn't test it yet, i hope it works.